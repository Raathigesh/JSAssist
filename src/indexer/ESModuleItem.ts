export interface Marker {
    location: SymbolLocation;
    filePath: string;
    name: string;
    isFromDefaultImport?: boolean;
}

export interface SymbolLocation {
    start: {
        line: number;
        column: number;
    };
    end: {
        line: number;
        column: number;
    };
}

export default class ESModuleItem {
    public id: string = '';
    public name: string = '';
    public location?: SymbolLocation;
    public path: string = '';
    public kind: string = '';
    public markers: Marker[] = [];
    public isDefaultExport: boolean = false;
}
