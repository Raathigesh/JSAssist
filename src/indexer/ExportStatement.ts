export default interface ExportStatement {
  path: string;
  specifiers: {
    exported: string;
    local: string;
  }[];
}
