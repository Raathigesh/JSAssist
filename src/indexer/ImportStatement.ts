export default class ImportStatement {
  public path: string = "";
  public defaultImportName: string | null = null;
  public namedImports: {
    name: string;
    aliasName?: string;
  }[] = [];
}
