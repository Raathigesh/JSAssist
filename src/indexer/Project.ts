export default interface Project {
  root: string;
  pathAlias: { [alias: string]: string };
}
