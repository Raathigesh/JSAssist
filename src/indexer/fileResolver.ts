import { isAbsolute, resolve, dirname, sep, join, normalize } from "path";
import { existsSync, statSync, lstatSync } from "fs";
const supportedExtensions = ["js", "jsx", "ts", "tsx"];

export function tryResolvingWithIndexOrExtension(originalPath: string) {
  for (const extension of supportedExtensions) {
    const pathWithExt = `${originalPath}.${extension}`;
    if (existsSync(pathWithExt)) {
      return pathWithExt;
    }
  }

  for (const extension of supportedExtensions) {
    const pathWithExt = join(originalPath, `index.${extension}`);
    if (existsSync(pathWithExt)) {
      return pathWithExt;
    }
  }

  return null;
}

export function getAliasPathForAbsolutePath(
  root: string,
  absolutePath: string,
  pathAliasMap: { [alias: string]: string }
) {
  let substitutedPath = normalize(absolutePath);
  Object.entries(pathAliasMap).forEach(([alias, path]) => {
    const normalizedPath = normalize(join(root, path));
    if (absolutePath.startsWith(normalizedPath)) {
      substitutedPath = absolutePath.replace(normalizedPath, alias);
    }
    const regexSep = sep === "\\" ? "\\\\" : sep;
    substitutedPath = substitutedPath.replace(new RegExp(regexSep, "g"), "/");
    if (substitutedPath.endsWith("index.js")) {
      substitutedPath = substitutedPath.replace(`${sep}index.js`, "");
    }
  });
  return substitutedPath;
}

export function findAbsoluteFilePathWhichExists(
  root: string,
  currentDirectory: string,
  originalPath: string,
  pathAliasMap: { [alias: string]: string }
) {
  // when a path is absolute
  if (isAbsolute(originalPath)) {
    if (existsSync(originalPath) && !lstatSync(originalPath).isDirectory()) {
      return originalPath;
    }

    const withIndex = tryResolvingWithIndexOrExtension(originalPath);
    if (withIndex) {
      return withIndex;
    }
  } else {
    // path is relative
    const absolutePath = resolve(currentDirectory, originalPath);
    if (existsSync(absolutePath)) {
      return absolutePath;
    } else {
      const withIndex = tryResolvingWithIndexOrExtension(absolutePath);
      if (withIndex) {
        return withIndex;
      }

      let substitutedPath = originalPath;
      Object.entries(pathAliasMap).forEach(([alias, path]) => {
        if (originalPath.startsWith(`${alias}/`)) {
          substitutedPath = originalPath.replace(alias, path);
          substitutedPath = resolve(root, substitutedPath);
        }
      });

      const subWithIndex = tryResolvingWithIndexOrExtension(substitutedPath);
      if (subWithIndex) {
        return subWithIndex;
      }
    }
  }

  throw Error("path does not exist");
}
