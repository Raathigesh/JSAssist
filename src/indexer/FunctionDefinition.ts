import ESModuleItem from "./ESModuleItem";

export default class FunctionDefinition extends ESModuleItem {}
