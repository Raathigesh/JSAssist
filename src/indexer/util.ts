import { FileType } from "indexer/FileType";
import { sep } from "path";

export const getFileType = (path: string): FileType => {
  const extension = path.split(".").pop();
  if (extension === "js") {
    return "JS";
  } else if (extension === "ts") {
    return "TS";
  } else if (extension === "tsx") {
    return "TSX";
  } else if (extension === "jsx") {
    return "JSX";
  }
  return "UNSUPPORTED";
};

export function santizePath(root: string, path?: string = "") {
  return path
    .replace(root, "")
    .split(sep)
    .join("/");
}
