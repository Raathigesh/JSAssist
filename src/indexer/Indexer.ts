const recursiveReadDir = require("recursive-readdir");
const fuzzysort = require("fuzzysort");
import { promisify } from "util";
import { Service } from "typedi";
import Project from "./Project";
import SourceFile, { ParseFailure } from "./SourceFile";
import { getFileType } from "indexer/util";
import ESModuleItem, { Marker } from "./ESModuleItem";
import { readFile, Stats } from "fs";
import { findAbsoluteFilePathWhichExists } from "./fileResolver";
import { dirname, join } from "path";
import { run } from "./WorkerRunner";

@Service()
export default class Indexer {
  public files: { [path: string]: SourceFile } = {};
  public status: "none" | "indexed" | "indexing" = "none";
  public project: Project | undefined;
  public totalFiles: number = 0;
  public indexedFileCount: number = 0;
  public failures: ParseFailure[] = [];
  public workerJSFile: string | undefined;

  public async parse(project: Project) {
    this.status = "indexing";
    this.project = project;
    this.failures = [];

    const files = await this.readProjectFiles(project.root);
    this.totalFiles = files.length;

    const supportedFiles = files.filter(
      filePath => getFileType(filePath) !== "UNSUPPORTED"
    );

    this.indexedFileCount = 0;

    return this.relaxedIndexer(supportedFiles);
  }

  public async relaxedIndexer(filesToIndex: string[]) {
    if (filesToIndex.length === 0) {
      this.status = "indexed";
      return;
    }

    if (this.project) {
      await run(
        {
          files: filesToIndex,
          pathAlias: this.project?.pathAlias,
          root: this.project?.root
        },
        (file: SourceFile) => {
          this.indexedFileCount += 1;
          this.files[file.path] = file;
        },
        this.workerJSFile
      );
      this.status = "indexed";
    }
  }

  public indexFile(path: string) {
    if (!this.project) {
      return;
    }

    const sourceFile = new SourceFile();
    this.files[path] = sourceFile;
    sourceFile.parse(path, this.project.pathAlias, this.project.root);
  }

  public search(query: string, queryType: string) {
    try {
      let searchQuery = query;
      const queryTypes = queryType.split(":");

      let results: ESModuleItem[] = [];
      Object.entries(this.files).forEach(([, file]) => {
        file.symbols.forEach(symbol => {
          if (
            queryType === "" ||
            (queryTypes.includes("func") &&
              symbol.kind === "FunctionDeclaration") ||
            (queryTypes.includes("type") && symbol.kind === "TypeAlias") ||
            (queryTypes.includes("var") &&
              symbol.kind === "VariableDeclaration") ||
            (queryTypes.includes("class") && symbol.kind === "ClassDeclaration")
          ) {
            results.push({
              ...symbol,
              path: file.path
            });
          }
        });
      });

      const filteredResults = fuzzysort.go(searchQuery, results, {
        key: "name",
        limit: 100
      });

      return filteredResults;
    } catch (e) {
      return [];
    }
  }

  public searchFile(query: string) {
    const filteredResults = fuzzysort.go(query, Object.keys(this.files), {
      limit: 100
    });
    return filteredResults;
  }

  public getSymbolWithMarkers(path: string, name: string) {
    const shouldGetDefaultExport = name === "@@DEFAULT_EXPORT@@";
    const file = this.files[path];

    if (shouldGetDefaultExport) {
      return file.symbols.find(symbol => symbol.isDefaultExport === true);
    }

    const symbol = file.symbols.find(symbol => symbol.name === name);

    if (!symbol && this.project) {
      const reExportedSymbol = this.getReExportedSymbol(name, path, file);
      if (reExportedSymbol) {
        return reExportedSymbol;
      }
    }
    return symbol;
  }

  public getSymbolsForPath(path: string) {
    const file = this.files[path];
    if (!file) {
      return [];
    }

    return file.symbols;
  }

  public findReference(name: string, path: string) {
    const markers: Marker[] = [];
    Object.values(this.files).forEach(file => {
      const symbols = file.symbols;
      symbols.forEach(symbol => {
        symbol.markers.forEach(marker => {
          if (
            marker.filePath === path &&
            marker.name === name &&
            symbol.location
          ) {
            markers.push({
              name: symbol.name,
              filePath: symbol.path,
              location: symbol.location
            });
          }
        });
      });
    });
    return markers;
  }

  public async getCode(path: string, name: string) {
    const shouldGetDefaultExport = name === "@@DEFAULT_EXPORT@@";
    const file = this.files[path];
    if (file) {
      let symbol: ESModuleItem | null | undefined;

      if (shouldGetDefaultExport) {
        symbol = file.symbols.find(symbol => symbol.isDefaultExport === true);
      }

      if (!symbol) {
        symbol = file.symbols.find(symbol => symbol.name === name);
      }

      if (!symbol || !symbol.location) {
        symbol = this.getReExportedSymbol(name, path, file);
        if (!symbol || !symbol.location) {
          return "";
        }
      }

      const content = await promisify(readFile)(symbol.path);
      const code = content.toString();
      const lines = code.split("\n");

      const {
        start = { line: 0, column: 0 },
        end = { line: 0, column: 0 }
      } = symbol.location;

      const results: string[] = [];

      lines.forEach((line, index) => {
        if (index >= start.line - 1 && index <= end.line - 1) {
          results.push(line);
        }
      });

      return results.reduce((acc, line) => {
        return acc + line + "\n";
      }, "");
    }
  }

  private async readProjectFiles(root: string) {
    if (this.project) {
      const indexableDirectories = this.project.directories.map(directory =>
        join(this.project?.root || "", directory)
      );

      function ignoreFunc(path: string, stats: Stats) {
        if (stats.isDirectory()) {
          return (
            !indexableDirectories.some(indexableDirectory =>
              indexableDirectory.includes(path)
            ) &&
            !indexableDirectories.some(indexableDirectory =>
              path.includes(indexableDirectory)
            )
          );
        }

        return (
          indexableDirectories.find(dir => path.includes(dir)) === undefined ||
          path.includes("node_modules")
        );
      }

      return new Promise<string[]>((resolve, reject) => {
        recursiveReadDir(root, [ignoreFunc], (err: Error, files: string[]) => {
          if (err) {
            reject(err);
          } else {
            resolve(files);
          }
        });
      });
    }

    return [];
  }

  private getReExportedSymbol(
    name: string,
    path: string,
    file: SourceFile
  ): ESModuleItem | null | undefined {
    if (!this.project) {
      return null;
    }

    let actualSymbolName = name;
    const exportStatement = file.exportStatements.find(s => {
      const specifier = s.specifiers.find(
        specifier => specifier.exported === name
      );
      if (specifier) {
        actualSymbolName = specifier?.local;
      }
      return specifier;
    });

    if (exportStatement) {
      const pathOfTheFile = exportStatement.path;
      const absolutePath = findAbsoluteFilePathWhichExists(
        this.project?.root,
        dirname(path),
        pathOfTheFile,
        this.project.pathAlias
      );
      const actualFile = this.files[absolutePath];
      const symbolFromActualFile = actualFile.symbols.find(
        symbol => symbol.name === actualSymbolName
      );

      if (symbolFromActualFile) {
        return symbolFromActualFile;
      }

      return this.getReExportedSymbol(name, absolutePath, actualFile);
    }
  }
}
