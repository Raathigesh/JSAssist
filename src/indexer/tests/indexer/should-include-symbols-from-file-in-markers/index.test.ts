import { resolve, sep, dirname } from 'path';
import Indexer from 'indexer/Indexer';
import Project from 'indexer/Project';
import { santizePath } from 'indexer/util';
import { Marker } from 'indexer/ESModuleItem';

const waitForIndexer = () =>
    new Promise(resolve => {
        setTimeout(() => {
            resolve();
        }, 2000);
    });

describe('Indexer', () => {
    it('should have symbols from the file in the marker', async () => {
        const project: Project = {
            root: resolve(__dirname, './project'),
            pathAlias: {},
            directories: ['.'],
            excludedDirectories: [],
        };
        const indexer = new Indexer();
        indexer.workerJSFile = '../../out/indexer/Worker.js';
        await indexer.parse(project);
        await waitForIndexer();

        const functionA = indexer.getSymbolWithMarkers(
            resolve(__dirname, './project/a.js'),
            'functionA'
        );

        expect({
            ...functionA,
            path: santizePath(project.root, functionA?.path),
            markers: functionA?.markers.map((marker: Marker) => ({
                ...marker,
                filePath: santizePath(project.root, marker.filePath),
            })),
        }).toMatchSnapshot();
    });
});
