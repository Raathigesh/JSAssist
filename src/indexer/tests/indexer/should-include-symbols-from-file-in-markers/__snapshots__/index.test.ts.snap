// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indexer should have symbols from the file in the marker 1`] = `
Object {
  "id": "ZnVuY3Rpb25BOkQ6XHByb2plY3RzXGluc2lnaHRcc3JjXGluZGV4ZXJcdGVzdHNcaW5kZXhlclxzaG91bGQtaW5jbHVkZS1zeW1ib2xzLWZyb20tZmlsZS1pbi1tYXJrZXJzXHByb2plY3RcYS5qcw==",
  "kind": "FunctionDeclaration",
  "location": SourceLocation {
    "end": Position {
      "column": 1,
      "line": 4,
    },
    "start": Position {
      "column": 7,
      "line": 2,
    },
  },
  "markers": Array [
    Object {
      "filePath": "/a.js",
      "location": Object {
        "end": Object {
          "column": 21,
          "line": 1,
        },
        "start": Object {
          "column": 12,
          "line": 1,
        },
      },
      "name": "functionB",
    },
  ],
  "name": "functionA",
  "path": "/a.js",
}
`;
